app/code/Magento/Catalog/Controller/Adminhtml/Product/Validate.php

```diff
@@ -149,6 +184,45 @@ public function execute()
         return $this->resultJsonFactory->create()->setData($response);
     }
 
+    /**
+     * Validates Url Key uniqueness.
+     *
+     * @param ProductModel $product
+     * @throws UrlAlreadyExistsException
+     */
+    private function validateUrlKeyUniqueness(ProductModel $product): void
+    {
+        if (!$product->getUrlKey()) {
+            $urlKey = $this->productUrlPathGenerator->getUrlKey($product);
+            $product->setUrlKey($urlKey);
+        }
+
+        $stores = $this->storeManager->getStores();
+
+        foreach ($stores as $store) {
+            if (!in_array($store->getWebsiteId(), $product->getWebsiteIds())) {
+                continue;
+            }
+
+            $urlPath = $this->productUrlPathGenerator->getUrlPathWithSuffix($product, $store->getId());
+
+            $urlRewrite = $this->urlFinder->findOneByData([
+                UrlRewrite::REQUEST_PATH => $urlPath,
+                UrlRewrite::ENTITY_TYPE => ProductUrlRewriteGenerator::ENTITY_TYPE,
+                UrlRewrite::STORE_ID => $store->getId()
+            ]);
```

Comment:
To reduce the impact of this, I believe we should use `findAllByData` and fetch all the data for all stores at once outside the foreach. As it is right now the time complexity of such would grow linear with the numbers of stores, where it can be just O(1).